src/traversing/findFilter.js

```diff
@@ -15,24 +15,29 @@ function winnow( elements, qualifier, not ) {
 		return jQuery.grep( elements, function( elem, i ) {
 			return !!qualifier.call( elem, i, elem ) !== not;
 		} );
-
 	}
 
+	// Single element
 	if ( qualifier.nodeType ) {
 		return jQuery.grep( elements, function( elem ) {
 			return ( elem === qualifier ) !== not;
 		} );
-
 	}
 
-	if ( typeof qualifier === "string" ) {
-		if ( risSimple.test( qualifier ) ) {
-			return jQuery.filter( qualifier, elements, not );
-		}
+	// Arraylike of elements (jQuery, arguments, Array)
+	if ( typeof qualifier !== "string" ) {
+		return jQuery.grep( elements, function( elem ) {
+			return ( indexOf.call( qualifier, elem ) > -1 ) !== not;
+		} );
+	}
 
-		qualifier = jQuery.filter( qualifier, elements );
+	// Simple selector that can be filtered directly, removing non-Elements
+	if ( risSimple.test( qualifier ) ) {
```

Comment:
It looks like a performance optimization for saving the O(n²) operations of invoking `indexOf` in a loop. And I'm guessing it only applies to simple selectors because `:not(non-simple selector)` is invalid CSS3 and therefore gets rejected by `querySelectorAll`, but if so that's irrelevant now because `jQuery.filter` only uses native methods when passed a single element. I'm in favor of using `return jQuery.filter( qualifier, elements, not )` for _all_ string `qualifier`s.